compiler/rustc_parse/src/lexer/tokentrees.rs

```diff
@@ -191,18 +183,50 @@ impl<'a> TokenTreesReader<'a> {
                 //     bar(baz(
                 // }  // Incorrect delimiter but matches the earlier `{`
                 if !self.diag_info.open_braces.iter().any(|&(b, _)| b == close_delim) {
-                    self.token = self.string_reader.next_token().0;
+                    self.bump(false).1
+                } else {
+                    // The choice of value here doesn't matter.
+                    Spacing::Alone
                 }
             }
             token::Eof => {
                 // Silently recover, the EOF token will be seen again
                 // and an error emitted then. Thus we don't pop from
-                // self.open_braces here.
+                // self.open_braces here. The choice of spacing value here
+                // doesn't matter.
+                Spacing::Alone
             }
             _ => unreachable!(),
-        }
+        };
+
+        let spacing = DelimSpacing::new(open_spacing, close_spacing);
 
-        Ok(TokenTree::Delimited(delim_span, open_delim, tts))
+        Ok(TokenTree::Delimited(delim_span, spacing, open_delim, tts))
+    }
+
+    // Move on to the next token, returning the current token and its spacing.
+    // Will glue adjacent single-char tokens together if `glue` is set.
+    fn bump(&mut self, glue: bool) -> (Token, Spacing) {
+        let (this_spacing, next_tok) = loop {
+            let (next_tok, is_next_tok_preceded_by_whitespace) = self.string_reader.next_token();
+
+            if is_next_tok_preceded_by_whitespace {
+                break (Spacing::Alone, next_tok);
+            } else if glue && let Some(glued) = self.token.glue(&next_tok) {
+                self.token = glued;
+            } else {
+                let this_spacing = if next_tok.is_punct() {
+                    Spacing::Joint
+                } else if next_tok.kind == token::Eof {
+                    Spacing::Alone
```

Comment:
Does this exception fixes cases like this?
```
// EOF immediately after string, so spacing is joint-hidden
"string"
```
The string token is then put into a token stream before an identifier.
```
"string"suffix
```
After pretty-printing and re-lexing the lexical structure is broken, `"string"suffix` is interpreted as 1 token instead of 2.

`Spacing::Alone` for EOF doesn't address the root issue here.
"string" can still get joint spacing in cases like `"string"+` and be printed as `"string"suffix` later.

This case needs a special case in pretty-printer.
Then special treatment for EOF in this file won't be necessary.
That's not an urgent issue though, it can be addressed separately later.

Reply:
The one crate for which this was a problem is described in [this comment above](https://github.com/rust-lang/rust/pull/114571#issuecomment-1813550126):

----

### DarumaDocker.reactor-wasm

Uses wasmedge_bindgen macro from wasmedge-bindgen-macro-0.1.13.

Weird one... this:
```rust
   let gen = quote! {
        #[no_mangle]
        pub unsafe extern "C" fn #func_ident(params_pointer: *mut u32, params_count: i32) {         
        ...
   }
```
somehow becomes this:
```rust
#[no_mangle] pub unsafe extern "C"fn run(params_pointer : * mut u32, params_count : i32) {
```
The lack of space after "C" is a legitimate problem. Hmm.

Somehow, Eof is involved. If we change things so that tokens followed by
Eof have Alone spacing instead of JointHidden spacing, that fixes the
problem. I don't understand why.

----

So it is a string suffix issue like you suggested. A follow-up sounds good, it only affected one crate in all the crater runs.